### PR TITLE
kernelci.org: tsc: add section about TSC rules

### PR DESCRIPTION
**Pull request by gctucker**
Add a section to detail the TSC rules in terms of votes and members. This is based on the project charter document and adds more context as to how this should work in practice.

A sub-section should be created as a follow-up to archive the results of all the votes carried out by the TSC.

Signed-off-by: Guillaume Tucker <guillaume.tucker@collabora.com>

### CODE REVIEW

**Comment by aliceinwire** [comment]
* make some form of positive contribution to the project
is too much generic for been considered a point to follow and could be cause of discussions about what is considered a "positive contribution" and not quantifiable.
The part about what active member are expected to do is also a big vague. I would prefer to rephrase as active member are expected to do at list one of the following action in one year:
* attend meetings (monthly)
* participate in votes
* Commit/Review to the project

**Comment by aliceinwire** [comment]
`Most decisions are reached by consensus and don't require a vote.` is not true "any" decision need a vote and need to be registered for transparency. consensus based without vote is not transparent and is not quantifiable. who/what consensus is this is also undefinable as there is no record of such process.
Voting should be defined. Vote process should use the majority vote of the current kernelci TSC members. we have a poll application in the Jitsi video chat let's use it or we can use a voting bot on IRC for better traceability. either way any decision need a vote and to be traced.

**Comment by aliceinwire** [comment]
`Typically, major contributors eventually become members` is too much abstract.
for example `Anyone can ask to be a member of kernelci TSC, by sending their own manifesto to the kernelci tsc mailing list, each member approval will be decided by a vote of the current TSC members.` would be more clear.

**Comment by aliceinwire** [comment]
`Votes done in a live meeting require 50% of the TSC to be
voting and a majority among the members attending.  Votes done by email or
other deferred channel require a majority among all of the TSC members.`
 such vote should be equal so both need the majority among all of the TSC members. there shall be no difference if is done live meeting or by email.

**Comment by nuclearcat** [comment]
> `Most decisions are reached by consensus and don't require a vote.` is not true "any" decision need a vote and need to be registered for transparency. consensus based without vote is not transparent and is not quantifiable. who/what consensus is this is also undefinable as there is no record of such process.

We need to define what exactly require mandatory voting, otherwise we will be drowning in formal voting on every minor decision. For example we already have a need for members attention, PR reviews, and in my experience there are few people who have time to do them, and even fewer people want to delve into the details of PR. If we add the need for frequent voting, things will get even slower.
For example recently i was not sure if i can introduce changes in mongodb configuration for memory management and upgrade of VM. Me and @mgalka weren't sure if this issue needed to be discussed at TSC, as it is important to some extent.
On the other hand, when production just completely stopped working, it left us no choice.

My suggestion, it might be worth doing this:
1) List of actions that require mandatory voting
2) In other cases, go through the formal notification of significant changes (maillist or IRC), (and in case of urgent changes, if they need to be done right now - make them with the possibility of cancellation/reversal, if that possible). If any member of the TSC has objections (during reasonable term) and voice them - then an action requires a vote.
3)List of things that definitely dont need voting? Like formal code changes, require only PR review.

**Comment by gctucker** [comment]
> `Most decisions are reached by consensus and don't require a vote.` is not true "any" decision need a vote and need to be registered for transparency. consensus based without vote is not transparent and is not quantifiable. who/what consensus is this is also undefinable as there is no record of such process. Voting should be defined. Vote process should use the majority vote of the current kernelci TSC members. we have a poll application in the Jitsi video chat let's use it or we can use a voting bot on IRC for better traceability or email. either way any decision need a vote and to be traced.

This is based on the LF project charter, quoting:

> 3.a While the Project aims to operate as a consensus based community, if any TSC
decision requires a vote to move the Project forward, the voting members of the
TSC will vote on a one vote per voting member basis.

**Comment by gctucker** [comment]
> `Typically, major contributors eventually become members` is too much abstract. for example `Anyone can ask to be a member of kernelci TSC, by sending their own manifesto to the kernelci tsc mailing list, each member approval will be decided by a majority vote of the current TSC members.` would be more clear.
> 
> Also I would prefer to remove the fact that `Current members of the TSC may suggest new people to be added. ` as this could make that only people in the TSC members circle became TSC member and degrading diversity in the TSC members.

How would adding and removing members would work if it's not something TSC members bring to the agenda for the TSC to vote?

If anyone from the community said either on a public channel or directly to a TSC member that they were willing to become part of the TSC then that would also be added to a TSC meeting agenda.  I don't think it's a closed system, just a process as only TSC members vote by definition.

**Comment by gctucker** [comment]
> `Votes done in a live meeting require 50% of the TSC to be voting and a majority among the members attending. Votes done by email or other deferred channel require a majority among all of the TSC members.` such vote should be equal so both need the majority among all of the TSC members. there shall be no difference if is done live meeting or by email. The problem that some TSC members are not joining the live meeting is our problem that we should improve for making them joining the meeting more easily and that should not change how we are voting.

This is also based on the LF project charter, so changing it would be more difficult.  We could say we never do vote in live meetings to enforce absolute 50% majority to be required.  Quoting 3.c:

> decisions by vote at a meeting require a majority vote of those in attendance, provided quorum is met. Decisions made
by electronic vote without a meeting require a majority vote of all voting members of the TSC.

**Comment by gctucker** [comment]
> The problem that some TSC members are not joining the live meeting is our problem that we should improve for making them joining the meeting more easily and that should not change how we are voting.

We've already done that by moving the day and time so you could join ;) (and others in APAC)  It's hard to have everyone attend a meeting with 17h of timezone difference.  Right now, it's only an inconvenient time for @khilman I believe and he said he was OK with not attending.

About the majority for voting, in practice we can definitely do a vote during a meeting and then share the result and topic on the kernelci-tsc mailing list so those who couldn't attend can also contribute their thoughts.  In the end, what's really important is to make the best decisions for the project.  The processes are important to ensure we have a minimum in place to function but we can always adjust things with some degree of flexibility.  Say, if the TSC voted in favour of some change, then we shared that on kernelci-tsc and someone who wasn't in the meeting replies with a very valid reason against that change which would cause the TSC members to change their minds then we could have another round of discussion and potentially another vote.  Or if a decision was made based on consensus and one person who was not present was against it then we could do the same and reconsider things or have a vote.

**Comment by gctucker** [comment]
About keeping track of things, vote results can be archived very precisely so that's useful.  For decisions made by consensus, we have the meeting minutes shared document.  We could extract a summary from each TSC meeting and publish that too.

**Comment by aliceinwire** [comment]
> > The problem that some TSC members are not joining the live meeting is our problem that we should improve for making them joining the meeting more easily and that should not change how we are voting.
> 
> We've already done that by moving the day and time so you could join ;) (and others in APAC) It's hard to have everyone attend a meeting with 17h of timezone difference. Right now, it's only an inconvenient time for @khilman I believe and he said he was OK with not attending.

I'm not talking only about the time. I'm talking about making the TSC enough useful for make members willing to join it and have something that they can do with the TSC that is useful also for the members. 

> 
> About the majority for voting, in practice we can definitely do a vote during a meeting and then share the result and topic on the kernelci-tsc mailing list so those who couldn't attend can also contribute their thoughts. In the end, what's really important is to make the best decisions for the project. The processes are important to ensure we have a minimum in place to function but we can always adjust things with some degree of flexibility. Say, if the TSC voted in favour of some change, then we shared that on kernelci-tsc and someone who wasn't in the meeting replies with a very valid reason against that change which would cause the TSC members to change their minds then we could have another round of discussion and potentially another vote. Or if a decision was made based on consensus and one person who was not present was against it then we could do the same and reconsider things or have a vote.

I'm strongly against having votes by some sort of consensus as is something that is not clearly defined and subjective by the people you are talking with also this clearly open the possibility of few persons taking over the organization.
Vote need to be done and need to be the majority of the TSC members for discussions relevant to TSC, with no exception.

**Comment by gctucker** [comment]
> I'm strongly against having votes by some sort of consensus

You're mixing things up, consensus is when there is no vote.  For example, moving sysadmin notifications to the kernelci-sysadmin@groups.io address didn't require a vote as nobody was against the idea and we all agreed it was a good thing to do.  Most technical decisions are just like this, with facts and logic we can find the best way forward.

**Comment by gctucker** [comment]
> I'm not talking only about the time. I'm talking about making the TSC enough useful for make members willing to join it and have something that they can do with the TSC that is useful also for the members.

So what do you suggest to make more people want to contribute to the project and eventually join the TSC?

**Comment by gctucker** [comment]
> bypass the fact that someone need to be choice from a member of the TSC

That's not what I meant nor what I wrote in this PR.  I'm saying that adding and removing members of the TSC should be done by a vote, and votes can only be proposed by existing members of the TSC by definition.  So one member of the TSC will have to add to the agenda that a vote is needed to add (or remove) someone from the TSC, regardless of how that came to be (email, IRC, discussion...).

**Comment by aliceinwire** [comment]
> > I'm strongly against having votes by some sort of consensus
> 
> You're mixing things up, consensus is when there is no vote. For example, moving sysadmin notifications to the [kernelci-sysadmin@groups.io](mailto:kernelci-sysadmin@groups.io) address didn't require a vote as nobody was against the idea and we all agreed it was a good thing to do. Most technical decisions are just like this, with facts and logic we can find the best way forward.

Yes that is not good. who decide there is consensus? consensus is not quantifiable or traceable. what I wrote still stand. I actually would have preferred there was a vote on that. like I would have preferred there was a vote on the introduction of kernelci-api and other technical parts that was concern of the TSC that never got a vote. This would help to keep from having one organization making most decision on the TSC problem without any vote or having one person deciding from some supposed consensus that mostly there isn't.

**Comment by aliceinwire** [comment]
> > bypass the fact that someone need to be choice from a member of the TSC
> 
> That's not what I meant nor what I wrote in this PR. I'm saying that adding and removing members of the TSC should be done by a vote, and votes can only be proposed by existing members of the TSC by definition. So one member of the TSC will have to add to the agenda that a vote is needed to add (or remove) someone from the TSC, regardless of how that came to be (email, IRC, discussion...).

you are misreading my arguments. I'm saying that TSC member need to be voted by TSC. just they don't need to be "selected" for been voted from TSC members, anyone willing to join the TSC that wrote a manifesto to the TSC email should be selected for having a vote of the TSC members for becoming a TSC member.

**Comment by aliceinwire** [comment]
> > I'm not talking only about the time. I'm talking about making the TSC enough useful for make members willing to join it and have something that they can do with the TSC that is useful also for the members.
> 
> So what do you suggest to make more people want to contribute to the project and eventually join the TSC?

I'm saying that if people don't join the TSC meeting is not others people problems but is a TSC problem. For example having the possibility to vote on TSC problem and not just deciding by some fake consensus is one of the reason that someone would be willing to join TSC.

**Comment by aliceinwire** [comment]
> > I'm strongly against having votes by some sort of consensus
> 
> You're mixing things up, consensus is when there is no vote. For example, moving sysadmin notifications to the [kernelci-sysadmin@groups.io](mailto:kernelci-sysadmin@groups.io) address didn't require a vote as nobody was against the idea and we all agreed it was a good thing to do. Most technical decisions are just like this, with facts and logic we can find the best way forward.

How do you know that nobody was against such idea without having a vote? I could actually have been against the idea and actually I was surprised to see such item in the TSC agenda already worked on, without any previous comment or discussion on TSC.

**Comment by gctucker** [comment]
> > > I'm strongly against having votes by some sort of consensus
> > 
> > 
> > You're mixing things up, consensus is when there is no vote. For example, moving sysadmin notifications to the [kernelci-sysadmin@groups.io](mailto:kernelci-sysadmin@groups.io) address didn't require a vote as nobody was against the idea and we all agreed it was a good thing to do. Most technical decisions are just like this, with facts and logic we can find the best way forward.
> 
> Yes that is not good. who decide there is consensus? consensus is not quantifiable or traceable. what I wrote still stand. I actually would have preferred there was a vote on that. like I would have preferred there was a vote on the introduction of kernelci-pipeline and other technical parts that was concern of the TSC that never got a vote. This would help to keep from having one organization making most decision on the TSC problem without any vote or having one person deciding from some supposed consensus that mostly there isn't.

Feel free to add things to the TSC meeting agenda and propose a vote, or bring things up on kernelci-tsc@groups.io.

About the example I picked, I asked during the meeting and everyone thought it was fine to move the notifications.  One topic I'll add for the next TSC meeting is whether this is OK or whether people think it would be better to vote on more technical things.

**Review by nuclearcat — APPROVED** [review]
LGTM, just i wonder what will happen if less than 50% of TSC members keep attending, i guess it will block voting process for quite long time. But this is more theoretical question, than concern.

**Comment by gctucker** [comment]
> LGTM, just i wonder what will happen if less than 50% of TSC members keep attending, i guess it will block voting process for quite long time. But this is more theoretical question, than concern.

Good point, it would probably be a sign that the people either aren't able to attend or are losing interest.  The particular issues behind it would need to be addressed, but I don't think we can have rules to prevent such issues.  At least the rules mean that without enough participation we can't make decisions any more, which seems like a good thing.  It also provides some framework and stability to the project which should hopefully help with increasing people's participation in the long term.

**Comment by khilman** [comment]
> Right now, it's only an inconvenient time for @khilman I believe and he said he was OK with not attending.

I was OK with this, but not anymore.  I resonate with the frustrations of @aliceinwire.  Without being able to regularly attend TSC meetings, it is difficult to follow important changes/decisions/priorities etc.   I would like to be able to attend meetings, at least occasionally.

**Comment by gctucker** [comment]
@khilman Sure, it'd be great to all attend the same meeting including yourself.  How do you suggest we do that?

With 17h difference between PST and JST (or 7h across the Pacific) it can't be practical for everyone to be in the same meeting.  Having multiple TSC meetings is not great either as it makes it harder for everyone to attend.

Maybe we can make sure some agenda is in place the day before the TSC meeting and any decision / votes made during the meeting can then be brought up on the mailing list?

Also I think it's important to clarify the scope of this PR, which is merely defining general rules about how the TSC operates.  We can always change how meetings and votes are done in practice within this framework, and adjust the framework as we go.  Is there anything in this PR you don't agree with?  If so, please add comments about what you would like to change.  It seems like we all ought to be discussing how to run the TSC in greater detail on the mailing list or in meetings rather than here.

**Review by khilman — APPROVED** [review]

**Comment by khilman** [comment]
For the specifics of this PR, I'm fine with it since like you said, it just documents the current state of things.  I just reviewed/approved it.

For the meetings, I'm not sure how to resolve it.  I don't think 2 meetings are great either.  My proposal would be to remain focused on the time slot where most active users can attend.  The current 11h00 UTC seems to be working ok for this.  However, I would request that occasionally (monthly?  every other month?) that we do a shifted one that US pacific folks could attend.

File: kernelci.org/content/en/docs/org/tsc.md
Comment (aliceinwire) [inline kernelci.org/content/en/docs/org/tsc.md:49]: > Most decisions are reached by consensus and don't require a vote.

I'm against this statement. 
We need to decide what require a vote and what doesn't require a vote in more detail. what if someone delete KernelCI GitHub ? will that need a vote or will be just "consensus"? 
KernelCI TSC should always need a vote and be completely transparent and recorded about discussion and decisions, if some decision can be accepted by consensus is not a topic of TSC but of maintainers or the weekly meeting. As quoted from the KernelCI TSC docs <The Technical Steering Committee (TSC) is a team of people who are responsible for keeping the KernelCI project in a good shape and driving its development.> for preserving the part of "team of people who are responsible" every TSC member need a vote on the topic discussed by TSC.

**Comment by aliceinwire** [comment]
> 



> > > > I'm strongly against having votes by some sort of consensus
> > > 
> > > 
> > > You're mixing things up, consensus is when there is no vote. For example, moving sysadmin notifications to the [kernelci-sysadmin@groups.io](mailto:kernelci-sysadmin@groups.io) address didn't require a vote as nobody was against the idea and we all agreed it was a good thing to do. Most technical decisions are just like this, with facts and logic we can find the best way forward.
> > 
> > 
> > Yes that is not good. who decide there is consensus? consensus is not quantifiable or traceable. what I wrote still stand. I actually would have preferred there was a vote on that. like I would have preferred there was a vote on the introduction of kernelci-pipeline and other technical parts that was concern of the TSC that never got a vote. This would help to keep from having one organization making most decision on the TSC problem without any vote or having one person deciding from some supposed consensus that mostly there isn't.
> 
> Feel free to add things to the TSC meeting agenda and propose a vote, or bring things up on [kernelci-tsc@groups.io](mailto:kernelci-tsc@groups.io).
> 
> About the example I picked, I asked during the meeting and everyone thought it was fine to move the notifications. One topic I'll add for the next TSC meeting is whether this is OK or whether people think it would be better to vote on more technical things.

This is the problem with consensus, how you can be sure that everyone is OK without a vote? and without votes been tracked?

**Comment by aliceinwire** [comment]
> > > > bypass the fact that someone need to be choice from a member of the TSC
> > > 
> > > 
> > > That's not what I meant nor what I wrote in this PR. I'm saying that adding and removing members of the TSC should be done by a vote, and votes can only be proposed by existing members of the TSC by definition. So one member of the TSC will have to add to the agenda that a vote is needed to add (or remove) someone from the TSC, regardless of how that came to be (email, IRC, discussion...).
> > 
> > 
> > you are misreading my arguments. I'm saying that TSC member need to be voted by TSC. just they don't need to be "selected" for been voted from TSC members, anyone willing to join the TSC that wrote a manifesto to the TSC email should be selected for having a vote of the TSC members for becoming a TSC member.
> 
> I think we agree on the principle but there's still some misunderstanding. Here's what I'm saying, let me know where you see a problem:
> 
>     * Adding or removing TSC members should be done by a TSC vote
> 
>     * TSC votes are carried out by TSC members during a TSC meeting or on [kernelci-tsc@groups.io](mailto:kernelci-tsc@groups.io)
> 
>     * TSC votes are supervised by the TSC chair to ensure the process is followed
> 
>     * As a result, regardless of how the suggestion to add or remove someone from the TSC came to be, it needs to go through the formal TSC voting process which is done by current members of the TSC.

I agree with this, my request was just to be sure that everyone can candidate themself by email without the necessity of a sponsor from one of the TSC members. Decision is still of course done by a TSC vote.

**Comment by aliceinwire** [comment]
> 



> LGTM, just i wonder what will happen if less than 50% of TSC members keep attending, i guess it will block voting process for quite long time. But this is more theoretical question, than concern.

I'm also worried about this and that is why I would prefer to have vote with the majority of all members in any case. also if the vote is done by video meetings.

**Comment by aliceinwire** [comment]
> LGTM, just i wonder what will happen if less than 50% of TSC members keep attending, i guess it will block voting process for quite long time. But this is more theoretical question, than concern.

If is some discussion that need to be moved on, you can move the voting process to email so that everyone can asynchronously vote.

**Comment by aliceinwire** [comment]
> @aliceinwire You've pointed out some things you disagree with, but in order to resolve the issue please make a clear proposal of how you would like the TSC rules to be defined. Also please keep in mind that this PR is mostly defining how things are currently working without changing any rules, and it's largely derived from the Linux Foundation project charter document which is the same for all projects. Changing that would require approval from the LF, e.g. definition of TSC vote and quorum etc.

@gctucker  This are the comments I could come out with now. for making a clear patch correction proposal I need more time as explained during the web meeting.

File: kernelci.org/content/en/docs/org/tsc.md
Comment (gctucker) [reply kernelci.org/content/en/docs/org/tsc.md:49]: If someone deletes a project on GitHub on purpose and without discussing it with anyone first, then it's obviously wrong and the person would most likely be removed from the TSC (with a vote...) if they're a member and have their admin permissions removed from GitHub.

Consensus for the TSC means the idea needs to be added to a meeting's agenda or on the kernelci-tsc mailing list, or at least on some chat channel e.g. Slack or IRC and others have a chance to say what they think.  It doesn't mean "free for all".  And as I've explained before already, if a consensus decision is made during a TSC meeting but one member could not attend and disagrees then it's always possible to restart the discussion.  We could have a rule that decisions should always have a date for when changes become effective, to leave some time for consideration and others to think about it.

**Comment by gctucker** [comment]
@aliceinwire OK I think it would be better to have this merged tomorrow and you can send a PR next week or whenever you have time to propose any changes.  In the meantime I'll start addressing the things mentioned here already:
* adjusting TSC meeting times for everyone to attend
* consider recording the meetings
* adding a section for archiving TSC decisions
* adding a rule about providing a date for when TSC decisions become effective

**Comment by gctucker** [comment]
Adding some clarity about what is expected from a TSC member to be "active" would also help, I'll think about this as well.

**Comment by aliceinwire** [comment]
> Adding some clarity about what is expected from a TSC member to be "active" would also help, I'll think about this as well.

that is the list wrote under.

File: kernelci.org/content/en/docs/org/tsc.md
Comment (aliceinwire) [reply kernelci.org/content/en/docs/org/tsc.md:49]: yes but you cannot know who voted for a decision. what if someone join a TSC meeting and maybe there are only three people and they decide to remove KernelCI, by consensus such thing is allowed and is even not tracked. I think that the added step about voting should help improve the transparency of the KernelCI TSC.
One example is how for example Gentoo manage technical discussion and decisions: https://projects.gentoo.org/council/meeting-logs/20220814.txt

Comment (gctucker) [reply kernelci.org/content/en/docs/org/tsc.md:49]: To be honest, if someone wanted to sabotage KernelCI they wouldn't even bother bringing it up as a discussion topic and there wouldn't be anything close to a consensus.  Or if they made it look like a consensus decision then it would obviously be addressed by the board and the offenders would most likely be removed from the TSC or have their admin rights removed etc. (as explained before)  So that's a fake issue, the real thing is we need to trust the people we're giving admin rights to with secure credentials (2FA) etc.  We also need recovery mechanisms in case some kind of disaster happens, voluntarily or not.

Comment (aliceinwire) [reply kernelci.org/content/en/docs/org/tsc.md:49]: That's not always so much obvious, like in the example I wrote. as that are only examples. Please take in consideration how other organizations are managing technical discussion and decisions.

Comment (aliceinwire) [reply kernelci.org/content/en/docs/org/tsc.md:49]: > the real thing is we need to trust the people we're giving admin rights to with secure credentials (2FA) etc. We also need recovery mechanisms in case some kind of disaster happens, voluntarily or not.

This is a issue but off topic in this pull request

Comment (gctucker) [reply kernelci.org/content/en/docs/org/tsc.md:49]: @aliceinwire I guess this is the rules followed by the Gentoo council?
https://wiki.gentoo.org/wiki/Project:Council

**Comment by gctucker** [comment]
One thing to keep in mind is that this PR is not the end of defining how the TSC operates but the beginning. There was nearly no documentation before, we just relied on the project technical charter provided by the LF and had some assumed rules based on how things have been done until now.

Writing all this down is the first step.  Then anyone can submit suggestions, create new PRs and the TSC can vote to change things and make the rules evolve.  One thing we can't easily do is change the LF project charter, so we should only consider doing this if there was a really critical issue.  It would require approval from the LF somehow, I'm not sure right now of all the implications this may have.

Another thing to do is look at how other LF projects operate, see how they've set things up with their TSC etc.   We can take some inspiration from that and also from other open source projects in general although every project has some particularities.  So we'll still need to tailor our rules in order to match our KernelCI LF project.

**Comment by gctucker** [comment]
Alright, merging this now so I can then send some follow-ups to address the things mentioned in this discussion.

**Pull request merged**
